contracts/LibSubmarineSimple.sol

```diff
@@ -0,0 +1,195 @@
+     * @notice Function called by the user to reveal the session.
+     * @dev warning Must be called within 256 blocks of the commit transaction to obtain the correct blockhash.
+     * @param _commitBlockNumber Block number in which the commit tx was included.
+     * @param _unlockAmount Value (i.e. how much money was sent) included in the commit tx.
+     * @param _embeddedDAppData optional Data passed embedded within the unlock tx. This should probably be null
+     * @param _witness Witness "secret" we committed to
+     * @param _unlockGasPrice Gas price of the unlock TX
+     * @param _unlockGasLimit Gas limit of the unlock TX
+     * @param _unlockTXHash full hash of the signed unlock transaction
+     * @param _proofBlob the Proof blob that gets passed to ethprove to verify merkle tree inclusion in a prior block.
+     */
+    function reveal(
+        uint32 _commitBlockNumber,
+        bytes _embeddedDAppData,
+        uint256 _unlockAmount,
+        bytes32 _witness,
+        uint256 _unlockGasPrice,
+        uint256 _unlockGasLimit,
+        bytes32 _unlockTXHash,
+        bytes _proofBlob
+    ) public {
+        bytes32 commitBlockHash = blockhash(_commitBlockNumber);
+        // fullCommit = (addressA + addressC + aux(sendAmount) + dappData + w + aux(gasPrice) + aux(gasLimit))
+
+        bytes32 commitId = getCommitId(
+            msg.sender,
+            address(this),
+            _unlockAmount,
+            _embeddedDAppData,
+            _witness,
+            _unlockGasPrice,
+            _unlockGasLimit
+        );
+
+        require(commitData[commitId].revealed == false, "The tx is already revealed");
+        require(commitBlockHash != 0x0, "Commit Block is too old to retreive block hash (more than 256 blocks), or does not exist");
+        require(block.number - _commitBlockNumber > commitPeriodLength, "You must wait long enough to allow the committing period to end before revealing");
+        ProvenTransaction memory proven_tx;
+        // Commented out for theoretical gas savings
+        (proven_tx.result, /* index */, proven_tx.nonce, /* gasprice */, /* startgas */, proven_tx.to, proven_tx.value, /* data */, /* v */ , /* r */, /* s */, proven_tx.is_contract_creation ) = txProof(commitBlockHash, _proofBlob);
+
+        // TX_PROOF_RESULT_PRESENT = 1;
+        // TX_PROOF_RESULT_ABSENT = 2;
+        require(proven_tx.result == 1, "The proof is invalid");
+        require(proven_tx.value >= _unlockAmount);
+        require(proven_tx.nonce == 0);
+        require(proven_tx.is_contract_creation == false);
+
+        address submarine = ecrecover(
+            _unlockTXHash,
+            vee,
+            keccak256(abi.encodePacked(commitId, byte(1))),
+            keccak256(abi.encodePacked(commitId, byte(0)))
+        );
+
+        require(keccak256(abi.encodePacked(proven_tx.to)) == keccak256(abi.encodePacked(submarine)), "The proven address should match the revealed address, or the txhash/witness is wrong.");
```

Comment:
Why do we compare the hash of the submarine address? Would it be cheaper to use `require (address(proven_tx.to) == submarine)`

Reply:
nevermind - the alternative is using assembly and there is barely any improvement.